api/agent/drivers/docker/docker.go

```diff
@@ -344,30 +344,52 @@ func dockerMsg(derr *docker.Error) string {
 // Run executes the docker container. If task runs, drivers.RunResult will be returned. If something fails outside the task (ie: Docker), it will return error.
 // The docker driver will attempt to cast the task to a Auther. If that succeeds, private image support is available. See the Auther interface for how to implement this.
 func (drv *DockerDriver) run(ctx context.Context, container string, task drivers.ContainerTask) (drivers.WaitResult, error) {
+
 	mwOut, mwErr := task.Logger()
+	attachSuccess := make(chan struct{})
+	stopSignal := make(chan struct{})
+	startWait := make(chan error, 1)
 
 	waiter, err := drv.docker.AttachToContainerNonBlocking(ctx, docker.AttachToContainerOptions{
-		Container: container, OutputStream: mwOut, ErrorStream: mwErr,
-		Stream: true, Stdout: true, Stderr: true,
-		Stdin: true, InputStream: task.Input()})
+		Success:      attachSuccess,
+		Container:    container,
+		OutputStream: mwOut,
+		ErrorStream:  mwErr,
+		Stream:       true,
+		Stdout:       true,
+		Stderr:       true,
+		Stdin:        true,
+		InputStream:  task.Input()})
+
 	if err != nil && ctx.Err() == nil {
 		// ignore if ctx has errored, rewrite status lay below
 		return nil, err
 	}
 
-	// we want to stop trying to collect stats when the container exits
-	// collectStats will stop when stopSignal is closed or ctx is cancelled
-	stopSignal := make(chan struct{})
-	go drv.collectStats(ctx, stopSignal, container, task)
+	// IMPORTANT: if attach NBIO did not return error above, then we know attachSuccess
+	// is going to be honored. See github.com/fsouza/go-dockerclient/client.go hijack()
+	if err == nil {
+		go func() {
+			<-attachSuccess
+			attachSuccess <- struct{}{}
 
-	err = drv.startTask(ctx, container)
-	if err != nil && ctx.Err() == nil {
-		// if there's just a timeout making the docker calls, drv.wait below will rewrite it to timeout
-		return nil, err
+			// we want to stop trying to collect stats when the container exits
+			// collectStats will stop when stopSignal is closed or ctx is cancelled
+			go drv.collectStats(ctx, stopSignal, container, task)
+
+			err = drv.startTask(ctx, container)
+			if err != nil && ctx.Err() == nil {
+				// if there's just a timeout making the docker calls, drv.wait below will rewrite it to timeout
+				startWait <- err
+			} else {
+				startWait <- nil
+			}
+		}()
```

Comment:
is there a reason to change the behavior here to return a container that may yet be started? to fix the attach bug the only code change we need appears to be (with error handling skipped):

```go
... : = drv.AttachToContainerNonBlocking(opts)

// wait to attach to container before we start it so that we do not miss output!
select {
case ctx.Done():
  return nil, ctx.Err()
case <-attachSuccess:
  attachSuccess <- struct{}{}
}
go drv.collectStats(...)

err = drv.startTask(...)
// handle err ...

return &waitResult{...}, nil
```

the above code ensures that we have attached to the container before it has started so that we do not miss any logs (the bug, as I understand it?) but doesn't require threading through start bugs all the way to the waiter, since `startTask` will give us a thumbs up or down on whether it started before returning.

Reply:
alternatively it seems like we could just use `AttachToContainer`? I think originally (or maybe b/c it is so) we had to use non blocking attach since attach would fail until the container was up, but my memory is foggy and the code here appears to do that and the tests pass.